sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java

```diff
@@ -37,102 +40,62 @@
  * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
  * at any time.
  */
-public final class AsynchronousMetricStorage<T> implements MetricStorage {
+public class AsynchronousMetricStorage<T, O> implements MetricStorage {
   private static final Logger logger = Logger.getLogger(AsynchronousMetricStorage.class.getName());
 
   private final ThrottlingLogger throttlingLogger = new ThrottlingLogger(logger);
-  private final MetricDescriptor metricDescriptor;
   private final ReentrantLock collectLock = new ReentrantLock();
-  private final AsyncAccumulator<T> asyncAccumulator;
+  private final ReentrantLock callbackLock = new ReentrantLock();
+  private final MetricDescriptor metricDescriptor;
   private final TemporalMetricStorage<T> storage;
-  private final Runnable metricUpdater;
+  private final AsyncAccumulator<T> accumulator;
+  private final O measurement;
 
-  /** Constructs asynchronous metric storage which stores nothing. */
-  public static MetricStorage empty() {
-    return EmptyMetricStorage.INSTANCE;
+  @GuardedBy("callbackLock")
```

Comment:
I think `CopyOnWriteArrayList` is commonly used to lock / copy on write instead of our current lock / copy on read. Though unfortunately to atomically check `isEmpty` and `forEach`, we may need to avoid that class and implement it ourselves, with a `volatile Consumer[] callbacks` and locks in add/remove callback (almost entirely to avoid `toArray`'s copy)

Reply:
Good point about `CopyOnWriteArrayList`.

I think we can get around having to do an atomic check of `isEmpty` and `forEach` with a bit of a trick. We can iterate through the callbacks and determine if callbacks was empty as a side effect of iteration:
```
  boolean hasCallbacks = false;
  for (Consumer<O> callback : callbacks) {
    hasCallbacks = true;
    callback.accept(measurement);
  }
  if (!hasCallbacks) {
    return EmptyMetricData.getInstance();
  }
```